### PR TITLE
[RLlib] Introduce FaultTolerantActorManager

### PR DESCRIPTION
**Pull request by gjoliver**
Signed-off-by: Jun Gong <jungong@anyscale.com>

## Why are these changes needed?

A remote actor manager that handles actor failures gracefully.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:128]: typo: a actor -> an actor

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:140]: typo: alwasy -> always (I highly recommend using copilot man!)

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:175]: use logger. warn or something?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:180]: why should this case not set the actor as un-healthy?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:206]: The behavior of exception pass through is extremely important and is missing from this documentation. User has to read the doc to understand its behavior.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:225]: should ignore worker failure be a parameter of the class or a parameter passed to `foreach_actor`?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:188]: What does this type `Callable[[T], T]` actually mean? `func(foo: int) -> int` is allowed but `func(foo: int) -> float` is not?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:249]: not clear what as only arg means?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:251]: nit: say "if True, applies `func` on known active actors only."

Comment (kouroshHakha) [reply rllib/utils/actor_manager.py:251]: I think you are interchanging the word `active` and `healthy` a lot in this PR. If they have the same meaning, can we just call it all either `active` or `healthy`?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:40]: Type is a `Mapping[ObjRef, int]` please clarify this here. I had to read the code to understand this.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:240]: The usage is not intuitive, so examples are needed. Because this returns the length of fired remote functions, but it's not clear what should I do next to fetch the results. Please add that to the preamble of the docstring along with some example usages.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:288]: should this be zero or none? also type is missing?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:300]: better wording: the index of the actors that were called

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:295]: variable missing from the function def

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:311]: instead of call can we use obj_ref as a more descriptive variable name?

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/tests/test_actor_manager.py:35]: Can add a doctsring on what this method does?  highlevel

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/tests/test_actor_manager.py:68]: typo: would not be aware

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/tests/test_actor_manager.py:72]: wait, I didn't understand where this 24 is coming from

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/tests/test_actor_manager.py:90]: again no clue why this is 33.

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/tests/test_actor_manager.py:72]: doesn't this depend on seed?

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:17]: nit: delete "states"

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:21]: Wait: Takes a `Callable[[Any], None]`, correct?

Meaning the lambda must have access to the actual object(!) behind the ActorHandle. Otherwise, it won't be able to do anything :)

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:25]: What is "async mode"?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:17]: Can you include the examples and their expected behavior for at least the use-cases that you are unittesting? After reading the unittests I am not sure if I can confidently explain what the expected behavior of this class is :) so some documentation here is indeed needed.

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:34]: Why two underscores?

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:49]: docstring

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:63]: docstring

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:68]: docstring

**Review by kouroshHakha — COMMENTED** [review]
Thanks, I have some important questions around the expected behavior and unittests.

File: rllib/utils/actor_manager.py
Comment (sven1977) [reply rllib/utils/actor_manager.py:62]: +1, no single-char var names.

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:138]: No tab/space indents in `Returns` descriptions. Messes up the sphinx-docs.

File: rllib/utils/actor_manager.py
Comment (sven1977) [reply rllib/utils/actor_manager.py:17]: So will this eventually completely replace the other Actor Manager that we already have and that's being used by e.g. Impala and APPO?

File: rllib/utils/actor_manager.py
Comment (sven1977) [reply rllib/utils/actor_manager.py:240]: Ah, async mode, got it. On a per-call basis.

File: rllib/utils/actor_manager.py
Comment (sven1977) [inline rllib/utils/actor_manager.py:16]: High level comment:

Looking at the API, I understand it mimics the one we already have for the non-flexible actor manager, but the more I'm thinking about this, the more I feel like this should replace the WorkerSet, whereas, right now, it's a parallel object to be handled by Algorithm besides the WorkerSet, even though, the WorkerSet and this manager share the same workers. This makes it very hard to understand, when I have to do something with my worker set (e.g. `add_worker`) or with my actor manager (also: e.g. `add_workers()`).

Then we could also make this manager automatically re-create the failed actors based on some flex-settings, like `min_num_actors`, and `max_num_actors` + our `ignore`/`restart` settings.

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:34]: yes. it's a python convention. one should absolutely not touch these members.
I am updating WorkerSet.remote_workers() to be ``__`` as well.
I think as a codebase, we have enjoyed too much functional programming convenience.

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:48]: I don't think we should make the API ambiguous? why can't the callers do the wrapping themselves?

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:180]: actor is still healthy. we need to differentiate actor failure from application failure.
let me add a comment about this.

File: rllib/utils/tests/test_actor_manager.py
Comment (gjoliver) [reply rllib/utils/tests/test_actor_manager.py:72]: yeah, we set the seed in actor.__init__()

File: rllib/utils/tests/test_actor_manager.py
Comment (gjoliver) [reply rllib/utils/tests/test_actor_manager.py:90]: same. all these unit tests work because we use fixed random seeds.

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:16]: maybe things will be clearer when I send the next PR where I update WorkerSet.

we can't make this a WorkerSet, because we have sets of actors that are not WorkerSet, namely Aggregation Workers, AlphaStar gradient workers, etc. So this needs to be a manager by itself that WorkerSet builds on top of.
We can extend this to support things like min_max_num_actors at some point, but I don't see the need right now. Things should work down to 1 actor.

Btw, I started with updating WorkerSet, exactly like your idea, and had to refactor this out at some point.

File: rllib/utils/tests/test_actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/tests/test_actor_manager.py:72]: would this not also depend on the machine / platform that runs these tests? how is the number of calls gonna be deterministic given the seed?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/actor_manager.py:16]: I just read the code for workerset yesterday for the first time in details. WorkerSet is also a manager of a set of RolloutWorker with specialized functionality for RolloutWorker API. I have a feeling that the "right" approach is to create a general fault tolerant ActorManager that will replace WorkerSet?

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:34]: this is super private. so private inherited classes don't even have access.
by coding style, this means "absolutely nobody touches these variables/functions".

File: rllib/utils/tests/test_actor_manager.py
Comment (gjoliver) [reply rllib/utils/tests/test_actor_manager.py:54]: done

**Review by gjoliver — COMMENTED** [review]
thanks for the thoughtful review. all done.
you comments also made me think that I should make the return type more explicit about the fact that we are returning:
1. results paired with the indices of the actors that handled them. and
2. both results and errors.
so I added a couple of utility classes to handle the specific return data and also makes it easy to skip errors, which is the most common response to errors that happen in this fault tolerant world.
please take another look.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/actor_manager.py:124]: maybe "replace" was a heavy word. You can either have workerset use Actor manager internally or it could be a sub-class of it with special functionality in place. To emphasize the shared functionality I think sub-classing is a better design if it can be pulled of.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/actor_manager.py:174]: In a subclass if you want to access a double underscored attribute python will yell at you.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [reply rllib/utils/actor_manager.py:239]: and if you del elements of each list, the list will re-index? how does that work for lists?

File: rllib/utils/actor_manager.py
Comment (sven1977) [reply rllib/utils/actor_manager.py:124]: Agree with @kouroshHakha , WorkerSet should probably be a sub-class of this new elastic ActorManager.

But we can do this later once this new class is more polished and tested. We should also remove the local worker from WorkerSet. It doesn't belong there.

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks for the great work @gjoliver and your thorough review @kouroshHakha .
Good to merge pending all test cases pass.

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:151]: What would happen if I didn't wait? would the fetch function return an empty list?

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:69]: I love this :)

File: rllib/utils/actor_manager.py
Comment (kouroshHakha) [inline rllib/utils/actor_manager.py:22]: wait where are these defined? `ResultOrError.Result()`

**Review by kouroshHakha — APPROVED** [review]
I love the new example and the new data structures you added. LGTM for now but needs to be tested thoroughly and proved to be robust until it stabilizes.

File: rllib/utils/actor_manager.py
Comment (sven1977) [reply rllib/utils/actor_manager.py:151]: yes

Comment (gjoliver) [reply rllib/utils/actor_manager.py:151]: yes

File: rllib/utils/actor_manager.py
Comment (gjoliver) [reply rllib/utils/actor_manager.py:69]: ack :)

**Comment by gjoliver** [comment]
> I love the new example and the new data structures you added. LGTM for now but needs to be tested thoroughly and proved to be robust until it stabilizes.

totally. so here is my plan:
1. get this merged.
2. move WorkerSet to wrap around this. But without any functional changes (call sites still access the underlying actors directly).
3. migrate algorithm to use the ``foreach_xxx`` utils. this is when I will run regression tests and preferably let Equilibre try it before merge.

**Comment by gjoliver** [comment]
Tests look all good. merging now.

**Pull request merged**
